src/amltk/scheduling/plugins/emissions_tracker_plugin.py

```diff
@@ -0,0 +1,118 @@
+from codecarbon import EmissionsTracker
+from typing import Callable, Any, ClassVar, TypeVar
+from typing_extensions import Self
+
+from amltk.scheduling.plugins.plugin import Plugin
+from amltk.scheduling.task import Task
+
+P = TypeVar("P")
+R = TypeVar("R")
+
+
+class _EmissionsTrackerWrapper:
+    """A wrapper around codecarbon package to measure emissions."""
+
+    def __init__(
+            self,
+            fn: Callable[[P], R],
+            task: Task,
+            *codecarbon_args: Any,
+            **codecarbon_kwargs: Any,
+    ):
+        """Initialize the wrapper.
+
+        Args:
+            fn: The function to wrap.
+            task: The task the function is being wrapped for.
+            *codecarbon_args: arguments to pass to
+                [`codecarbon.EmissionsTracker`][codecarbon.EmissionsTracker].
+            **codecarbon_kwargs: keyword arguments to pass to
+                [`codecarbon.EmissionsTracker`][codecarbon.EmissionsTracker].
+        """
+        super().__init__()
+        self.fn = fn
+        self.task = task
+        self.codecarbon_args = codecarbon_args
+        self.codecarbon_kwargs = codecarbon_kwargs
+
+    def __call__(self, *args: any, **kwargs: any) -> R:
```

Comment:
Small info on how class generics work:

If you make the class generic with `_EmissionsTrackerWrapper(Generic[P, R]):`, you would be saying that the class is generic with respect to what it gets in its `__init__` call, in this case `fn: Callable[P, R]`. This then means anything in the class you can access the `P` generic and `R` generic and it will specifically be the type `P`, which are a [`ParamSpec`](https://devdocs.io/python~3.10/library/typing#typing.ParamSpec) ("Parameter Spec") of what `fn` takes in and `R` as what `fn` returns.

Then you can say that `__call__` takes in the same parameter as `self.fn` takes in and returns the same thing that `self.fn` does, as you've already done with `-> R`.

```python
def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
    ...
```